internal/cri/server/restart.go

```
@@ -177,6 +177,15 @@ func (c *criService) recover(ctx context.Context) error {
 
 				return nil
 			}
+			defer func() {
+				if egoErr == nil || cntr.IO == nil {
+					return
+				}
+
+				if errClose := cntr.IO.Close(); errClose != nil {
+					log.G(ctx2).WithError(errClose).Errorf("Failed to close container io %q", cntr.ID)
```

hmm if the store add is successful .. but the name reservation fails.. then we need to call containerStore.Delete which will do the close for the console if IO is not nil.. 

really we need a way to know if we should call delete I think..

---

I suppose you could just do the close directly before the error return for the failed add below.. and then on add success do a defer for the delete before calling reserve..

---

another thing to consider is.. why does loadContainer not have it's own defer to close the console it created... hmm

---

> hmm if the store add is successful .. but the name reservation fails.. then we need to call containerStore.Delete which will do the close for the console if IO is not nil..

Good point! I think we should do reservation first and then add it into store, like what we do in pod/container creation. 

Since we have several goroutines to do that, we may be aware that if there are several containers with same name, first come (reservation) first serve.

---

done. when I review delete I find anoter bug may cause io failed to close. https://github.com/containerd/containerd/pull/11407/files

---

I review https://github.com/containerd/containerd/pull/11407  don't need to return an error.

---

I'm not sure whether is it necessary to delete conflict containers? or let k8s delete it? but k8s may forget to call delete.

---

If we should delete container we should stop and delete container.

---

Would you please make sure your change can be build first? the CI is failed.

For this iteration, I don't think we should delete container in this recovery because stop/delete could take long time, which impacts recovery process and causes kubelet is not ready. My suggestion was that

* do reservation for name first
* and then add container into store

```go
eg.Go(func() (egErr error) {
			cntr, err := c.loadContainer(ctx2, container)
			if err != nil {
				log.G(ctx2).
					WithError(err).
					WithField("container", container.ID()).
					Error("Failed to load container")

				return nil
			}
			log.G(ctx2).Debugf("Loaded container %+v", cntr)
                         defer func() {
                              if egErr != nil && cntr.IO != nil {
                                     cntr.IO.Close() // log error
                              }
                         }()
                        
			if err := c.containerNameIndex.Reserve(cntr.Name, cntr.ID); err != nil {
				return fmt.Errorf("failed to reserve container name %q: %w", cntr.Name, err)
			}
                        
			if err := c.containerStore.Add(cntr); err != nil {
				c.containerNameIndex.ReleaseByName(cntr.Name)
				return fmt.Errorf("failed to add container %q to store: %w", container.ID(), err)
			}
			return nil
		})
```

---

done